macro-pdfviewer-api/src/main/java/com/xwiki/pdfviewer/filter/MJSFilter.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.pdfviewer.filter;
+
+import java.lang.reflect.Field;
+
+import javax.inject.Singleton;
+
+import org.apache.tika.Tika;
+import org.apache.tika.config.TikaConfig;
+import org.apache.tika.mime.MediaType;
+import org.apache.tika.mime.MimeTypes;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.tika.internal.TikaUtils;
+
+/**
+ * Filter component for checking the mime type of .mjs files.
+ *
+ * @version $Id$
+ * @since 2.6
+ */
+@Component(roles = MJSFilter.class)
+@Singleton
+public class MJSFilter
+{
+    /**
+     * If the correct mime type for .mjs files is not set, uses reflection to modify the {@link TikaUtils} {@link Tika}
+     * object with the right mime type configuration for .mjs files.
+     */
+    public void maybeAddMJSMimeType()
+    {
+        try {
+            String mjsMimeType = TikaUtils.detect("test.mjs");
+            if (!mjsMimeType.equalsIgnoreCase("application/javascript") && !mjsMimeType.equalsIgnoreCase(
+                "text/javascript"))
+            {
+                Field tikaField = TikaUtils.class.getDeclaredField("tika");
+                tikaField.setAccessible(true);
+                TikaConfig tikaConfig = new TikaConfig(TikaUtils.class.getResource("/tika-config.xml"));
+                tikaConfig.getMimeRepository()
+                    .addPattern(MimeTypes.getDefaultMimeTypes().forName(MediaType.application("javascript").toString()),
+                        "*.mjs");
+
+                // Create a new Tika instance with an associated mime type for the "*.mjs" pattern.
+                Tika newTika = new Tika(tikaConfig);
+                tikaField.set(null, newTika);
```

Comment:
The problem with this solution is that it doesn't have a **fallback** in case:

* there is a security policy in place that forbids making the private ``tika`` field accessible and settings its value
* the Tika code is refactored and the private ``tika`` field is renamed or removed

AFAIK Tika already supports registering new mime types using https://tika.apache.org/2.9.2/parser_guide.html#Add_your_MIME-Type , but there are two problems:

* it needs a server restart
* ``custom-mimetypes.xml`` needs to be in the class-path accessible by Tika but since Tika is a **core** extension, it won't see files provided by installed extensions AFAIK (so it won't work to put this file in the PDF Viewer Macro jar). Instead, we need to provide the steps to manually activate this file for the wiki administrator.

Reply:
> custom-mimetypes.xml needs to be in the class-path accessible by Tika but since Tika is a core extension, it won't see files provided by installed extensions AFAIK (so it won't work to put this file in the PDF Viewer Macro jar). Instead, we need to provide the steps to manually activate this file for the wiki administrator.

Indeed, adding the `custom-mimetypes.xml` file to the PDF Viewer Macro jar is not helpful as it is not seen by Tika. I think the best course of action beside this solution would be to manually add the custom mime types file to the XWiki jar and restart the server. Maybe have a script that would automate the file change and a documentation that would explain how to manually do this if the application and the script fails?

Reply:
I think it's enough to:

* indicate in the logs what needs to be done, in case adding the mime type programmatically fails
* update the PDF Viewer Macro documentation to mention this case and provide the ``custom-mimetypes.xml`` file that needs to be put by an administrator in ``WEB-INF/classes/org/apache/tika/mime/`` (inside the XWiki WAR).